### PR TITLE
fix(semantic): find the nearest valid jsdoc comment

### PR DESCRIPTION
**Pull request by leaysgur**
Current implementation fails to find jsdoc comment in some cases.

Like,

```js
/** @typedef {{ x: 1 }} X */
/** @type {X} */
let x;
```

For `let x;` it attempts to match `@typedef` comments, but nothing but whitespace can exist between them, so nothing matches as a result.

- - -

BTW, in this case, how should we get the first `@typedef` comment? It does not belong to any node but valid usage.

### CODE REVIEW

File: crates/oxc_semantic/src/jsdoc/mod.rs [hunk 1]
```diff
@@ -17,7 +17,7 @@ pub struct JSDoc<'a> {
     docs: BTreeMap<Span, JSDocComment<'a>>,
 }
 
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, PartialEq, Eq)]
```
Comment (leaysgur) [inline crates/oxc_semantic/src/jsdoc/mod.rs:20]: This was added only for test, is it OK?

**Comment by leaysgur** [comment]
It seems that tests containing comments in their test case are failing..., I'll check it later.

File: crates/oxc_semantic/src/jsdoc/mod.rs (hunk 1 above)
Comment (Boshen) [reply crates/oxc_semantic/src/jsdoc/mod.rs:20]: Yes this is ok.

**Comment by Boshen** [comment]
> Besides Program, is there a larger AstKind that encompasses individuals, which may need to be excluded...?

It's easiest to work with a whitelist with unit tests added incrementally, otherwise you'll end up with unmanaged situations 😅 

Only excluding `Program` here may hurt performance a little bit because it'll check all AST nodes, which is a lot.

**Comment by leaysgur** [comment]
@Boshen Thank you for your comment!

I tried a bit and realized that I should avoid whitelists, as they are still redundant to manage. 😓 

What I want to do here is to extend this original code,

https://github.com/leaysgur/oxc/blob/7a6261403db9163cad6638365c1f1053608f1061/crates/oxc_semantic/src/jsdoc/builder.rs#L57-L62

to allow normal comments as well as white-space.

I think I'm stuck on this now, do you have any good ideas?

**Comment by leaysgur** [comment]
Closing once, I think I need more research and learning to cope with this. 🍃 

Current problems I found are https://github.com/oxc-project/oxc/issues/168#issuecomment-1900022366

**Pull request closed without merging**